Create EIP for frame-pointer-relative memory addressing (#3337)
ethereum/EIPs: EIPS/eip-3337.md

### `MLOADFP` opcode
A new opcode `MLOADFP` is introduced with value `0x5e`. This opcode acts in all ways identical to `MLOAD`, except that the value of the frame pointer is added to the address before loading data from memory. An attempt to load data from a negative address should be treated identically to an invalid opcode, consuming all gas and reverting the current execution context.

### `MSTOREFP` opcode
A new opcode `MSTOREFP` is introduced with value `0x5f`. This opcode acts in all ways identical to `MSTORE`, except that the value of the frame pointer is added to the address before storing data to memory. An attempt to store data to a negative address should be treated identically to an invalid opcode, consuming all gas and reverting the current execution context.

## Rationale
### Cost of new opcodes
The cost of the new opcodes `MLOADFP` and `MSTOREFP` reflects the cost of `MLOAD` and `MSTORE`. They are generally equivalent in cost with the exception of an extra addition operation, which imposes negligible cost.

The cost of the new opcodes `SETFP` and `GETFP` is based on other common low-cost opcodes such as `PUSH` and `POP`.

### Absence of `MSTORE8FP`

> Whether or not to include an `MSTORE8FP` opcode depends on the frequency with which this will be used, and the desire to minimise the size of the instruction set and to conserve opcodes for future use. If the community believes this will be a substantial asset, the EIP can be modified to include it.

Reviewer: Write EIPs targeting a future reader, which means not including commentary on potential changes to the EIP.
```suggestion
`MSTORE8FP` opcode was not included because it is expected that it would be used infrequently, and there is a desire to minimise the size of the instruction set and to conserve opcodes for future use.
```

## Backwards Compatibility
This EIP exclusively introduces new opcodes, and as a result should not impact any existing programs unless they operate under the assumption that these opcodes are undefined, which we believe will not be the case.

> ## Test Cases
> TBD
> 

Reviewer: ```suggestion
```
Not a required section, so I recommend just leaving it out until such time in the future as you fill it in.  If you want to leave this in as a reminder to yourself, that is also acceptable.

Likewise, before this EIP a frame-pointer relative store requires the following sequence of operations:
| Opcode    | Gas used |
|-----------|----------|
| `PUSHn x` | 3        |
| `PUSH1 0` | 3        |
| `MLOAD`   | 3        |
| `ADD`     | 3        |
| `MSTORE`  | 3        |

This consumes 15 gas and at least 7 bytes of bytecode. After this EIP, the equivalent sequence of operations is:

| Opcode    | Gas used |
|-----------|----------|
| `PUSHn x` | 3        |
| `MSTOREFP`| 3        |

Consuming only 6 gas and at least 3 bytes of bytecode, while once again only requiring EVM implementations to do one extra addition operation. The alternative of storing values on the stack requires 6 gas and 2 bytes of bytecode for the sequence `SWAPn POP`, making it no more efficient than memory storage.

## Specification
### Parameters

| Constant | Value |
| - | - |
| `FORK_BLOCK` | TBD |

For blocks where `block.number >= FORK_BLOCK`, the following changes apply.

### Frame pointer

> A new EVM internal state variable called the "frame pointer" is introduced. This is a signed integer that starts at 0.

Reviewer 2: Probably needs to be specified as 256-bit twos-complement for the sake of the GETFP and SETFP operations.

Reviewer: You may want to bring this up in the discussions-to to make sure it gets seen.  I agree that this needs more specificity!

Side note: We probably don't need 256-bits, likely 32-bits would be way more than enough?
